pkg/cloudprovider/providers/aws/aws.go

```diff
@@ -1264,30 +1263,85 @@ func (c *Cloud) getMountDevice(
 	// Check to see if this volume is already assigned a device on this machine
 	for mountDevice, mappingVolumeID := range deviceMappings {
 		if volumeID == mappingVolumeID {
-			if assign {
-				glog.Warningf("Got assignment call for already-assigned volume: %s@%s", mountDevice, mappingVolumeID)
-			}
 			return mountDevice, true, nil
 		}
 	}
+	return mountDevice(""), false, nil
+}
+
+// attachNewDevice attached a new disk to the node.
+func (c *Cloud) attachNewDevice(i *awsInstance, info *ec2.Instance, volumeID awsVolumeID) (mountDevice, error) {
+	// pick a new device name from pool of 52 devices (ba..bz,ca..cz). We are trying to prevent
+	// device name reuse as much as possible of recently freed devices. Reducing device name
+	// reuse on AWS helps prevent volume stuck in 'attaching' state bug.
+	chosen, deviceAllocator, alreadyAttached, err := c.allocateDiskDevice(i, info, volumeID)
+
+	if err != nil {
+		return "", err
+	}
+
+	glog.V(2).Infof("Assigned mount device %s -> volume %s", chosen, volumeID)
+	if !alreadyAttached {
+		attachResult, attachErr := c.performAwsAttach(chosen, volumeID, i)
+		if attachResult && deviceAllocator != nil {
+			// We only 'mark' device as taken when attach succeeds. We don't want failing attach
+			// calls to use device names.
+			deviceAllocator.MarkForUse(chosen)
```

Comment:
`deviceAllocator.MarkForUse` does not mark the device as taken, it just gives it the least priority when choosing a next device. MarkForUse does not sound good to me...